### DISCUSSION TITLE
ScriptReference/MonoBehaviour.OnValidate

### DISCUSSION THREAD

**Discussion by giscus** [Developer-Notes-Extension/Unity-Notes#36] [Notes]
# ScriptReference/MonoBehaviour.OnValidate

https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnValidate.html

**Comment by vertxxyz** [comment]
Do not cache variables with a null check in OnValidate unless they are global, as if you copy-paste component values they will not be contextual to your new instance.

**Reply by vertxxyz** [reply]
Also, if you want to use this to cache variables into serialize fields, I would avoid it. `OnValidate` is called *a lot*, and is a likely culprit contributing to editor slow-down.  
Instead, use [`Reset`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.Reset.html), which is only called when adding or resetting components. Avoid using `HideInInspector`, just allow users to see the reference and reconfigure if they'd like.

**Comment by docscomments** [comment]
**Comment by `@litefeel`:**  

This message can be sent from assets and instances.
